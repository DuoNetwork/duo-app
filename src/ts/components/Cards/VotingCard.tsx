import { IEsplanadeStates, IVotingData , Web3Wrapper} from '@finbook/duo-contract-wrapper';
import moment from 'moment';
import * as React from 'react';
import * as CST from 'ts/common/constants';
import { IAddresses } from 'ts/common/types';
import { esplanadeWrapper } from 'ts/common/wrappers';
import { SDivFlexCenter } from '../_styled';
import { SCard, SCardList, SCardTitle, SInput } from './_styled';

interface IProps {
	locale: string;
	states: IEsplanadeStates;
	votingData: IVotingData;
	account: string;
	coldAddressPool: IAddresses;
	moderator: string;
	candidate: string;
}

interface IState {
	candidateContractAddr: string;
	candidateContractAddrErr: string;
}

export default class EsplanadeVotingCard extends React.Component<IProps, IState> {
	constructor(props: IProps) {
		super(props);
		this.state = {
			candidateContractAddr: '',
			candidateContractAddrErr: ''
		};
	}
	private handleCandidateContractAddressChange = (address: string) =>
		this.setState({
			candidateContractAddr: address,
			candidateContractAddrErr: Web3Wrapper.checkAddress(address)
				? ''
				: 'Invalid Address'
		});

	public handleVote = (isVotingFor: boolean) =>
		esplanadeWrapper.vote(this.props.account, isVotingFor);

	public handleStartContractVoting = () => {
		if (this.state.candidateContractAddr && !this.state.candidateContractAddrErr)
			esplanadeWrapper.startContractVoting(
				this.props.account,
				this.state.candidateContractAddr
			);
	};

	public handleStartModeratorVoting = () =>
		esplanadeWrapper.startModeratorVoting(this.props.account);

	public handleTerminateVoting = () =>
		esplanadeWrapper.terminateContractVoting(this.props.account);

	public handleTerminateByTimeout = () => esplanadeWrapper.terminateByTimeout(this.props.account);

	public render() {
		const { locale, votingData, states, coldAddressPool, moderator, candidate } = this.props;
		const { candidateContractAddr, candidateContractAddrErr } = this.state;

		const isInVoting = states.votingStage !== 'NotStarted';
		const isModerator = this.props.account === moderator;
		const isInCold = Object.keys(coldAddressPool).includes(this.props.account);

		return (
			<SCard
				title={
					<SCardTitle>
						<a
							className="tag-content"
							href={
								'https://' +
								(__KOVAN__ ? 'kovan.' : '') +
								'etherscan.io/address/' +
								esplanadeWrapper.web3Wrapper.contractAddresses.MultiSigManagers[0]
									.address
							}
							target="_blank"
							style={{ color: 'white' }}
						>
							{CST.TH_VOTING.toUpperCase()}
						</a>
					</SCardTitle>
				}
				width={'1000px'}
				margin={'0 10px 0 10px'}
			>
				<SDivFlexCenter horizontal padding="0 10px">
					<SCardList noMargin>
						<div className="status-list-wrapper">
							<ul style={{ paddingTop: '10px' }}>
								<li>
									<span className="title">{CST.TH_VOTING_STAGE}</span>
									<span className="content">{states.votingStage}</span>
								</li>
								<li>
									<span className="title">{CST.TH_MODERATOR}</span>
									<span className="content">{moderator}</span>
									<button
										className={'form-button'}
										disabled={!isInCold}
										onClick={() => this.handleStartModeratorVoting()}
									>
										{CST.TH_START_MODERATOR_VOTING}
									</button>
								</li>
								<li>
									<span className="title">{CST.TH_CANDIDATE}</span>
									<span className="content">{candidate}</span>
									<button
										className={'form-button'}
										disabled={!isInCold}
										onClick={() => this.handleTerminateByTimeout()}
									>
										{CST.TH_TERMINATE_BY_TIME_OUT}
									</button>
								</li>
								<li>
									<span className="title">{CST.TH_STARTED}</span>
									<span className="content">
										{votingData.started
											? moment
													.utc(votingData.started)
													.format('YYYY-MM-DD HH:mm:SS')
											: 'never'}
									</span>
									<button
										className={'form-button'}
										disabled={!isModerator || states.votingStage !== 'Contract'}
										onClick={() => this.handleTerminateVoting()}
									>
										{CST.TH_TERMINATE_VOTING}
									</button>
								</li>
								<li>
									<span className="title">{CST.TH_VOTE_FOR}</span>
									<span className="content">{`${votingData.votedFor}/${
										votingData.totalVoters
									}`}</span>
									<button
										className={'form-button'}
										disabled={!isInVoting || !isInCold}
										onClick={() => this.handleVote(true)}
									>
										{CST.TH_VOTE_FOR}
									</button>
								</li>
								<li>
									<span className="title">{CST.TH_VOTE_AGAINST}</span>
									<span className="content">{`${votingData.votedAgainst}/${
										votingData.totalVoters
									}`}</span>
									<button
										className={'form-button'}
										disabled={!isInVoting || !isInCold}
										onClick={() => this.handleVote(false)}
									>
										{CST.TH_VOTE_AGAINST}
									</button>
								</li>
								<li className="no-bg">
									<span className="title">{CST.TH_START_CONTRACT_VOTING}</span>
									<SInput
										className={candidateContractAddrErr ? 'input-error' : ''}
										placeholder={CST.TT_INPUT_ADDR[locale]}
										width={'400px'}
										value={candidateContractAddr}
										onChange={e =>
											this.handleCandidateContractAddressChange(
												e.target.value
											)
										}
										small
									/>
									<button
										className={'form-button'}
										disabled={!isModerator}
										onClick={() => this.handleStartContractVoting()}
									>
										{CST.TH_START_CONTRACT_VOTING}
									</button>
								</li>
							</ul>
						</div>
					</SCardList>
				</SDivFlexCenter>
			</SCard>
		);
	}
}
