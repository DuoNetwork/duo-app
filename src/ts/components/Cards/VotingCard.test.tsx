// fix for @ledgerhq/hw-transport-u2f 4.28.0
import '@babel/polyfill';
import { shallow } from 'enzyme';
import * as React from 'react';
import VotingCard from './VotingCard';

describe('EsplanadeCard Test', () => {
	describe('EsplanadeCard', () => {
		const states = {
			isStarted: false,
			votingStage: 'NotStarted',
			poolSizes: {
				cold: 0,
				hot: 0,
				custodian: 0,
				otherContract: 0
			},
			operationCoolDown: 0,
			lastOperationTime: 0
		};
		const coldAddressPool = {
			address0: {
				balance: 10,
				index: 0
			},
			address1: {
				balance: 20,
				index: 1
			},
			address2: {
				balance: 30,
				index: 2
			}
		};
		const votingData = {
			started: 0,
			votedFor: 1,
			votedAgainst: 2,
			totalVoters: 10
		}

		it('Test Snapshot', () => {
			const wrapper1 = shallow(
				<VotingCard
					account={'0x0'}
					coldAddressPool={coldAddressPool}
					states={states}
					locale={'EN'}
					votingData={votingData}
					moderator={'moderator'}
					candidate={'candidate'}
				/>
			);
			expect(wrapper1).toMatchSnapshot();
			const wrapper2 = shallow(
				<VotingCard
					account={'moderator'}
					coldAddressPool={coldAddressPool}
					states={states}
					locale={'EN'}
					votingData={votingData}
					moderator={'moderator'}
					candidate={'candidate'}
				/>
			);
			expect(wrapper2).toMatchSnapshot();
		});
	});
});
