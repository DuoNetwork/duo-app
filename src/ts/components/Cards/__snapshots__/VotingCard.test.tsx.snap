// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EsplanadeCard Test EsplanadeCard Test Snapshot 1`] = `
<Styled(Card)
  margin="0 10px 0 10px"
  title={
    <styled.div>
      <a
        className="tag-content"
        href="https://kovan.etherscan.io/address/0xa43f6d00b3F2efe39012b56aaA04Eee58a0F5079"
        style={
          Object {
            "color": "white",
          }
        }
        target="_blank"
      >
        VOTING
      </a>
    </styled.div>
  }
  width="1000px"
>
  <styled.div
    horizontal={true}
    padding="0 10px"
  >
    <styled.div
      noMargin={true}
    >
      <div
        className="status-list-wrapper"
      >
        <ul
          style={
            Object {
              "paddingTop": "10px",
            }
          }
        >
          <li>
            <span
              className="title"
            >
              Voting Stage
            </span>
            <span
              className="content"
            >
              NotStarted
            </span>
          </li>
          <li>
            <span
              className="title"
            >
              Moderator
            </span>
            <span
              className="content"
            >
              moderator
            </span>
            <button
              className="form-button"
              disabled={true}
              onClick={[Function]}
            >
              Start Moderator Voting
            </button>
          </li>
          <li>
            <span
              className="title"
            >
              Candidate
            </span>
            <span
              className="content"
            >
              candidate
            </span>
            <button
              className="form-button"
              disabled={true}
              onClick={[Function]}
            >
              Terminate By Time Out
            </button>
          </li>
          <li>
            <span
              className="title"
            >
              Started
            </span>
            <span
              className="content"
            >
              never
            </span>
            <button
              className="form-button"
              disabled={true}
              onClick={[Function]}
            >
              Terminate Voting
            </button>
          </li>
          <li>
            <span
              className="title"
            >
              Vote For
            </span>
            <span
              className="content"
            >
              1/10
            </span>
            <button
              className="form-button"
              disabled={true}
              onClick={[Function]}
            >
              Vote For
            </button>
          </li>
          <li>
            <span
              className="title"
            >
              Vote Against
            </span>
            <span
              className="content"
            >
              2/10
            </span>
            <button
              className="form-button"
              disabled={true}
              onClick={[Function]}
            >
              Vote Against
            </button>
          </li>
          <li
            className="no-bg"
          >
            <span
              className="title"
            >
              Start Contract Voting
            </span>
            <styled.input
              className=""
              onChange={[Function]}
              placeholder="Please input address"
              small={true}
              value=""
              width="400px"
            />
            <button
              className="form-button"
              disabled={true}
              onClick={[Function]}
            >
              Start Contract Voting
            </button>
          </li>
        </ul>
      </div>
    </styled.div>
  </styled.div>
</Styled(Card)>
`;

exports[`EsplanadeCard Test EsplanadeCard Test Snapshot 2`] = `
<Styled(Card)
  margin="0 10px 0 10px"
  title={
    <styled.div>
      <a
        className="tag-content"
        href="https://kovan.etherscan.io/address/0xa43f6d00b3F2efe39012b56aaA04Eee58a0F5079"
        style={
          Object {
            "color": "white",
          }
        }
        target="_blank"
      >
        VOTING
      </a>
    </styled.div>
  }
  width="1000px"
>
  <styled.div
    horizontal={true}
    padding="0 10px"
  >
    <styled.div
      noMargin={true}
    >
      <div
        className="status-list-wrapper"
      >
        <ul
          style={
            Object {
              "paddingTop": "10px",
            }
          }
        >
          <li>
            <span
              className="title"
            >
              Voting Stage
            </span>
            <span
              className="content"
            >
              NotStarted
            </span>
          </li>
          <li>
            <span
              className="title"
            >
              Moderator
            </span>
            <span
              className="content"
            >
              moderator
            </span>
            <button
              className="form-button"
              disabled={true}
              onClick={[Function]}
            >
              Start Moderator Voting
            </button>
          </li>
          <li>
            <span
              className="title"
            >
              Candidate
            </span>
            <span
              className="content"
            >
              candidate
            </span>
            <button
              className="form-button"
              disabled={true}
              onClick={[Function]}
            >
              Terminate By Time Out
            </button>
          </li>
          <li>
            <span
              className="title"
            >
              Started
            </span>
            <span
              className="content"
            >
              never
            </span>
            <button
              className="form-button"
              disabled={true}
              onClick={[Function]}
            >
              Terminate Voting
            </button>
          </li>
          <li>
            <span
              className="title"
            >
              Vote For
            </span>
            <span
              className="content"
            >
              1/10
            </span>
            <button
              className="form-button"
              disabled={true}
              onClick={[Function]}
            >
              Vote For
            </button>
          </li>
          <li>
            <span
              className="title"
            >
              Vote Against
            </span>
            <span
              className="content"
            >
              2/10
            </span>
            <button
              className="form-button"
              disabled={true}
              onClick={[Function]}
            >
              Vote Against
            </button>
          </li>
          <li
            className="no-bg"
          >
            <span
              className="title"
            >
              Start Contract Voting
            </span>
            <styled.input
              className=""
              onChange={[Function]}
              placeholder="Please input address"
              small={true}
              value=""
              width="400px"
            />
            <button
              className="form-button"
              disabled={false}
              onClick={[Function]}
            >
              Start Contract Voting
            </button>
          </li>
        </ul>
      </div>
    </styled.div>
  </styled.div>
</Styled(Card)>
`;
